test/unit/org/apache/cassandra/utils/CassandraGenerators.java

```diff
@@ -0,0 +1,299 @@
+    {
+        Gen<AbstractType<?>> typeGen = getType();
+        switch (kind)
+        {
+            // partition and clustering keys require frozen types, so make sure all types generated will be frozen
+            // empty type is also not supported, so filter out
+            case PARTITION_KEY:
+            case CLUSTERING:
+                typeGen = Generators.filter(typeGen, t -> t != EmptyType.instance).map(t -> t.freeze());
+                break;
+        }
+        if (kind == ColumnMetadata.Kind.CLUSTERING)
+        {
+            // when working on a clustering column, add in reversed types periodically
+            typeGen = allowReversed(typeGen);
+        }
+        // filter for unique names
+        String str;
+        while (!createdColumnNames.add(str = IDENTIFIER_GEN.generate(rnd)))
+        {
+        }
+        ColumnIdentifier name = new ColumnIdentifier(str, true);
+        int position = !kind.isPrimaryKeyKind() ? -1 : (int) rnd.next(Constraint.between(0, 30));
+        return new ColumnMetadata(ks, table, name, typeGen.generate(rnd), position, kind);
+    }
+
+    public static Gen<ByteBuffer> partitionKeyDataGen(TableMetadata metadata)
+    {
+        ImmutableList<ColumnMetadata> columns = metadata.partitionKeyColumns();
+        assert !columns.isEmpty() : "Unable to find partition key columns";
+        if (columns.size() == 1)
+            return getTypeSupport(columns.get(0).type).bytesGen();
+        List<Gen<ByteBuffer>> columnGens = new ArrayList<>(columns.size());
+        for (ColumnMetadata cm : columns)
+            columnGens.add(getTypeSupport(cm.type).bytesGen());
+        return rnd -> {
+            ByteBuffer[] buffers = new ByteBuffer[columnGens.size()];
+            for (int i = 0; i < columnGens.size(); i++)
+                buffers[i] = columnGens.get(i).generate(rnd);
+            return CompositeType.build(buffers);
+        };
+    }
+
+    /**
+     * Hacky workaround to make sure different generic MessageOut types can be used for {@link #MESSAGE_GEN}.
+     */
+    private static final Gen<Message<?>> cast(Gen<? extends Message<?>> gen)
+    {
+        return (Gen<Message<?>>) gen;
+    }
+
+    /**
+     * Java's type inferrence with chaining doesn't work well, so this is used to inferr the root type early in cases
+     * where javac can't figure it out
+     */
+    private static <T> Gen<T> gen(Gen<T> fn)
+    {
+        return fn;
+    }
+
+    /**
+     * Converts the type to a CQL type.  This method special cases empty and UDTs so the string can be used in a create
+     * statement.
+     *
+     * Special cases
+     * <ul>
+     *     <li>empty - replaces with 'org.apache.cassandra.db.marshal.EmptyType'.  empty is the tostring of the type in
+     *     CQL but not allowed to create as empty, but fully qualified name is allowed</li>
+     *     <li>UserType - replaces with TupleType</li>
+     * </ul>
+     */
+    public static String toCqlType(AbstractType<?> type)
```

Comment:
I'm wondering if this method might be more at home in `SchemaCQLHelper`.

Reply:
will move, it does